### PR TITLE
Use package.json files array instead of .npmignore

### PR DESCRIPTION
**Pull request by jakebailey**
In #50327, I forgot to add `.eslintcache` to `.npmignore`, meaning depending on how the build is run, this file may have appeared in our output package. This was fixed in #50380.

Rather than needing to remember to add any new file to `.npmignore`, it's (in my opinion) a lot safer to instead explicitly declare which files we want to be shipping instead. You can do that by using `"files"` in `package.json`; the meanings are reversed to `.npmignore`, but you can still write ignores via the `!` prefix.

The result is the same as our current `.npmignore`, with a few exceptions:

- `lib/README.md` is not explicitly ignored. This was in our `.npmignore`, except that it never worked because npm _always_ includes files named README. See: https://unpkg.com/browse/typescript@4.7.4/lib/README.md
- `CopyrightNotice.txt`: This is not the same as our license; it's just a file that we prepend to our outputs to add a header and should really be in some build folder. LICENSE.txt is our license. (When we move to modules, I plan to use copyright headers and a lint rule, so this file will disappear anyway.)
- `CODE_OF_CONDUCT.md`, which we were required to add by policy in #27581, but doesn't need to be in our published package and was likely overlooked similarly to my `.eslintcache` file.

The downside is that if you're adding a new file and it's not in one of the expected locations, it won't be included. But, I think it's unlikely that we add files in different locations, compared to _very_ likely that we add more things like config files for tooling.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,6 +28,17 @@
     "engines": {
         "node": ">=4.2.0"
     },
+    "files": [
+        "./bin",
+        "./lib",
+        "!./lib/enu",
+        "./loc",
```
Comment (DanielRosenwasser) [inline package.json:35]: Should `loc` be shipped? What is it for?

Comment (jakebailey) [reply package.json:35]: I honestly have no clue; it's already in the package we publish. Maybe that's an oversight too!

File: package.json [hunk 2]
```diff
@@ -28,6 +28,17 @@
...
+        "./AUTHORS.md",
+        "./LICENSE.txt",
```
Comment (DanielRosenwasser) [inline package.json:37]: This already has it - but do we need `CopyrightNotice.txt`?

File: package.json [hunk 3]
```diff
@@ -28,6 +28,17 @@
     "engines": {
         "node": ">=4.2.0"
     },
+    "files": [
+        "./bin",
+        "./lib",
+        "!./lib/enu",
```
Comment (DanielRosenwasser) [inline package.json:34]: Do we have any idea what `enu` is?

File: package.json (hunk 1 above)
Comment (jakebailey) [reply package.json:35]: I did a quick by-hand bisect and loc was first added to the package in 3.7; I have yet to go look why.

File: package.json (hunk 2 above)
Comment (jakebailey) [reply package.json:37]: See my PR description; this file is a JS comment that the build prepends to JS files during build. I don't think it should be in the package.

Comment (DanielRosenwasser) [reply package.json:37]: You already mentioned this I suppose. Makes sense.

File: package.json (hunk 3 above)
Comment (jakebailey) [reply package.json:34]: I believe these are the XML files that the localization team consumes. Our localize task produces both the output we use as well as these files to give to them to use in some GUI program.

File: package.json (hunk 1 above)
Comment (DanielRosenwasser) [reply package.json:35]: I don't think it needs to be shipped. It is likely part of the configuration for our localization team. Maybe @sandersn, @csigs, or @pdostalek know.

File: package.json [hunk 4]
```diff
@@ -28,6 +28,17 @@
     "engines": {
         "node": ">=4.2.0"
     },
+    "files": [
+        "./bin",
```
Comment (jakebailey) [inline package.json:32]: I made these all `./`, but that's not actually required I think.

File: package.json (hunk 1 above)
Comment (jakebailey) [reply package.json:35]: Removing it shaves off 78 files and 2.7 MB (!) unpacked, so removing it would be great.

File: package.json (hunk 4 above)
Comment (jakebailey) [reply package.json:32]: (But, I'll probably leave it that way, as it mirrors the other paths we use in the file; see above.)

File: package.json (hunk 1 above)
Comment (jakebailey) [reply package.json:35]: In fact, this folder hasn't been touched in 3 years; does it still contain anything that's in use?

File: package.json (hunk 3 above)
Comment (jakebailey) [reply package.json:34]: Now, I'm not actually sure that this is copied in the `lkg` build; I will check.

(Personally, I think that these localization team things would be better suited for a different task, as we don't need it in our dev workflow day-to-day, but that's a topic for another PR.)

File: package.json (hunk 1 above)
Comment (jakebailey) [reply package.json:35]: After talking in standup, I'm going to send a PR that deletes this folder altogether from the repo.

**Review by DanielRosenwasser — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (DanielRosenwasser) [reply package.json:35]: My concern with that is a possible situation where some infra breaks and we stop getting localization updates. But the PR itself is a good place to discuss further.

Comment (jakebailey) [reply package.json:35]: Yeah, for now, this PR keeps the status quo except the limited number of exceptions I listed above.

**Pull request merged**